kernel/arch/riscv/process.c

```diff
@@ -64,29 +221,221 @@ int setup_sig_handler(struct task *ti,
                       ulong user_func,
                       int signum)
 {
-   NOT_IMPLEMENTED();
+   if (ti->nested_sig_handlers == 0) {
+
+      int rc;
+
+      if (sig_state == sig_pre_syscall)
+         r->a0 = (ulong) -EINTR;
+
+      if ((rc = save_regs_on_user_stack(ti, r)) < 0)
+         return rc;
+   }
+
+   r->sepc = user_func;
+   r->usersp -= SIG_HANDLER_ALIGN_ADJUST;
+   r->a0 = (ulong)signum;
```

Comment:
Can you move this code to signal.c and make it portable using the inline funcs for accessing `usersp`, `ip` and other typical register meta-names that are schared across architectures ?

Reply:
fixed.